contracts/GroupVoteLocker.sol

```diff
@@ -0,0 +1,331 @@
+            }
+
+            // Calculate the change in bias for the current epoch
+            int128 biasDelta = lastCheckpoint.slope *
+                SafeCast.toInt128(
+                    int256((iterativeTime - lastCheckpointTimestamp))
+                );
+
+            // The bias can be below 0
+            lastCheckpoint.bias = max(lastCheckpoint.bias - biasDelta, 0);
+            // The slope should never be below 0 but added for safety
+            lastCheckpoint.slope = max(lastCheckpoint.slope + slopeDelta, 0);
+            lastCheckpoint.ts = iterativeTime;
+            lastCheckpointTimestamp = iterativeTime;
+            lastCheckpoint.blk =
+                initialLastCheckpoint.blk +
+                ((blockSlope * (iterativeTime - initialLastCheckpoint.ts)) /
+                    1e18); // Scale back down
+
+            groupState.epoch += 1;
+
+            if (iterativeTime == block.timestamp) {
+                lastCheckpoint.blk = block.number;
+                // Adjust the last checkpoint for any delta from the user
+                lastCheckpoint.slope = max(
+                    lastCheckpoint.slope + userSlopeDelta,
+                    0
+                );
+                lastCheckpoint.bias = max(
+                    lastCheckpoint.bias + userBiasDelta,
+                    0
+                );
+                groupState.checkpoints.push(lastCheckpoint);
+                break;
+            } else {
+                groupState.checkpoints.push(lastCheckpoint);
+            }
+        }
+    }
+
+    /**
+     * @dev Get the total vote power from the group state at the latest block
+     * @param groupState group state to calculate voting power from
+     */
+    function totalVotePower(GroupVotePowerState storage groupState) internal view returns (uint256) {
+        if (groupState.checkpoints.length == 0) {
+            return 0;
+        }
+        Checkpoint memory lastCheckpoint = groupState.checkpoints[groupState.epoch];
+        return _votePowerAt(groupState, lastCheckpoint, block.timestamp);
+    }
+
+    /**
+     * @dev Get the total vote power from the group state at the specified _blockNumber
+     * @param groupState group state to calculate voting power from
+     * @param _blockNumber block number at which totalVotePower is returned
+     */
+    function totalVotePowerAt(GroupVotePowerState storage groupState, uint256 _blockNumber) internal view returns (uint256) {
+        require(_blockNumber <= block.number, "Block number is in the future");
+    
+        // Get most recent global Checkpoint to block
+        uint256 recentGroupStateEpoch = _findEpoch(
+            groupState.checkpoints,
+            _blockNumber,
+            groupState.epoch
+        );
+
+        if (recentGroupStateEpoch == 0) {
+            return 0;
+        }
+
+        Checkpoint memory checkpoint0 = groupState.checkpoints[recentGroupStateEpoch];
+
+        if (checkpoint0.blk > _blockNumber) {
```

Comment:
Should this be a require? _findEpoch is supposed to always return something before the timestamp, unless it's the 0 epoch, and we have check above for the zero epoch. Theoretically, hitting this conditional should be impossible?